### PR TITLE
feat(cni): add l3forward flow

### PR DESCRIPTION
**Pull request by qingwusunny**

### CODE REVIEW

File: pkg/monitor/helper.go [hunk 1]
```diff
@@ -54,3 +57,50 @@ func (fn ovsUpdateHandlerFunc) Stolen([]interface{}) {
...
+		return nil
+	}
+	return net.ParseIP(ipStr)
+}
+
+func getDriverNameFromInterface(row ovsdb.Row) string {
+	if row.Fields[InterfaceStatus] == nil {
+		return ""
+	}
+	status := row.Fields[InterfaceStatus].(ovsdb.OvsMap).GoMap
```
Comment (wu-cl) [inline pkg/monitor/helper.go:77]: add status check here, such as:
```
if item, ok := row.Fields["external_ids"].(ovsdb.OvsMap); ok {}
```

File: pkg/monitor/helper.go [hunk 2]
```diff
@@ -54,3 +57,50 @@ func (fn ovsUpdateHandlerFunc) Stolen([]interface{}) {
...
+
+func getMacStrFromInterface(row ovsdb.Row) (string, error) {
+	driver := getDriverNameFromInterface(row)
+	if driver == "" {
+		return "", fmt.Errorf("get interface driver failed, interface row: %+v", row)
+	}
+
+	var macStr string
+	if driver == VMNicDriver || driver == PodNicDriver {
+		externalIds := row.Fields["external_ids"].(ovsdb.OvsMap).GoMap
```
Comment (wu-cl) [inline pkg/monitor/helper.go:93]: ditto

File: pkg/monitor/ovsdb.go [hunk 3]
```diff
@@ -250,14 +253,12 @@ func (monitor *OVSDBMonitor) filterLocalEndpoint(rowupdate ovsdb.RowUpdate) (*da
...
-	if !ok {
-		macStr, _ = rowupdate.New.Fields["mac_in_use"].(string)
-	} else {
-		macStr, _ = newExternalIds[LocalEndpointIdentity].(string)
+	macStr, err := getMacStrFromInterface(rowupdate.New)
+	if err != nil {
+		klog.Infof("Get endpoint mac failed, error: %v", err)
+		return nil, nil
 	}
+	newExternalIds := rowupdate.New.Fields["external_ids"].(ovsdb.OvsMap).GoMap
```
Comment (wu-cl) [inline pkg/monitor/ovsdb.go:261]: ditto

File: pkg/monitor/ovsdb.go [hunk 4]
```diff
@@ -288,20 +288,19 @@ func (monitor *OVSDBMonitor) filterEndpointUpdated(rowupdate ovsdb.RowUpdate) (*
...
-	if !ok {
-		macStr, _ = rowupdate.New.Fields["mac_in_use"].(string)
-	} else {
-		macStr, _ = newExternalIds[LocalEndpointIdentity].(string)
+	macStr, err := getMacStrFromInterface(rowupdate.New)
+	if err != nil {
+		klog.Infof("Get endpoint mac failed, error: %v", err)
+		return nil
 	}
+	newExternalIds := rowupdate.New.Fields["external_ids"].(ovsdb.OvsMap).GoMap
```
Comment (wu-cl) [inline pkg/monitor/ovsdb.go:302]: ditto

File: pkg/agent/datapath/localBridge.go [hunk 5]
```diff
@@ -275,6 +254,8 @@ func (l *LocalBridge) BridgeInitCNI() {
 		sw := l.OfSwitch
 		l.cniConntrackCommitTable, _ = sw.NewTable(CNI_CT_COMMIT_TABLE)
 		l.cniConntrackRedirectTable, _ = sw.NewTable(CNI_CT_REDIRECT_TABLE)
+		// cni mode doesn't open mac learning
```
Comment (wu-cl) [inline pkg/agent/datapath/localBridge.go:257]: what about remove `arp-to-controller` flow

File: pkg/monitor/helper.go (hunk 1 above)
Comment (qingwusunny) [reply pkg/monitor/helper.go]: done

File: pkg/monitor/helper.go (hunk 2 above)
Comment (qingwusunny) [reply pkg/monitor/helper.go]: done

File: pkg/monitor/ovsdb.go (hunk 4 above)
Comment (qingwusunny) [reply pkg/monitor/ovsdb.go]: done

File: pkg/agent/datapath/localBridge.go (hunk 5 above)
Comment (qingwusunny) [reply pkg/agent/datapath/localBridge.go]: add a ip learning switch

File: pkg/monitor/ovsdb.go [hunk 6]
```diff
@@ -522,6 +517,7 @@ func (monitor *OVSDBMonitor) processOvsInterfaceUpdate(uuid string, rowupdate ov
 		InterfaceUUID: oldEndpoint.InterfaceUUID,
 		BridgeName:    oldEndpoint.BridgeName,
 		MacAddrStr:    oldEndpoint.MacAddrStr,
+		IPAddr:        oldEndpoint.IPAddr,
```
Comment (echkenluo) [inline pkg/monitor/ovsdb.go:520]: Is this safe ? IPAddr is implement as []byte.

Comment (qingwusunny) [reply pkg/monitor/ovsdb.go]: done

**Review by echkenluo — APPROVED** [review]

**Review by echkenluo — APPROVED** [review]

**Review by echkenluo — APPROVED** [review]

File: pkg/agent/rpcserver/cni.go [hunk 7]
```diff
@@ -172,6 +172,7 @@ func (s *CNIServer) CmdAdd(ctx context.Context, request *cnipb.CniRequest) (*cni
 		Name:      "pod-" + string(args.K8S_POD_NAME),
 		Namespace: string(args.K8S_POD_NAMESPACE),
 	})
+	externalID["attached-ipv4"] = result.IPs[0].Address.IP.String()
```
Comment (wu-cl) [inline pkg/agent/rpcserver/cni.go:175]: remove L168 comments, no arp learning any more

**Review by wu-cl — APPROVED** [review]

**Review by echkenluo — APPROVED** [review]

**Pull request merged**